src/ripple/protocol/impl/SOTemplate.cpp

```diff
@@ -28,18 +28,20 @@ void SOTemplate::push_back (SOElement const& r)
     //
     if (mIndex.empty ())
     {
-        // Unmapped indices will be set to -1
+        // Unmapped indices are initialized to -1
         //
         mIndex.resize (SField::getNumFields () + 1, -1);
     }
 
     // Make sure the field's index is in range
     //
-    assert (r.e_field.getNum () < mIndex.size ());
+    if (r.e_field.getNum() <= 0 || r.e_field.getNum() >= mIndex.size())
+        Throw<std::runtime_error> ("Invalid field index for SOTemplate.");
```

Comment:
I'm curious about your take on exceptions here. As near as I can tell, this code will be accessed by a couple of static instance constructors, but the items we pass are fixed/known - so won't ever throw? I realize we'd probably like a violation of these invariants to be caught at compile time, but that's tricky here. Is there a scenario outside of unit tests that would trigger an exception and will it be handled as we expect?

Reply:
That's a fair question.  Let me see if I can recreate what I was thinking when I made this change.

I believe my concern was that `assert`s only occur in a debug build.  If a bad `SField` is passed here then previously the bounds checking only occurred in a debug build.  In a release build we would index out of range, leading to unknown bad things.

Also, `SOTemplate` construction is relatively rare so, in terms of execution time, performing the validation in a release build adds very little aggregate runtime cost.  I think we can afford to do this validation in a release build.

It just increased my confidence a bit if we do the `SOTemplate` validation checking all the time, not just in a debug build.  And that's not something I get from an `assert`.  But I also didn't want to blindly shutdown a release-build of a server, which is why I did not use `LogicError`.

So I can't say I have clear evidence that this is a necessary change.  I can't point to a way that foreign code could cause an invalid `SOTemplate` to be constructed in a release build.  I simply preferred to put in a check that both can't be ignored and doesn't shutdown the server.

The consequence of the `throw` would be that the `SOTemplate` would still exist, but it would not contain all of the intended fields.  That's because fields are added by `SOTemplate::push_back()` not in the `SOTemplate` constructor.

That could be an excellent argument to add a constructor to the `SOTemplate` that takes a `std::initializer_list<SOElement>`.  I think if I did that I could remove `SOTemplate::push_back`.  That would lead to the consequence that if an invalid field were part of the `SOTemplate` the `SOTemplate` constructor would throw, so the entire `SOTemplate` with the invalid field would not even exist.

Let me know if you think adding the `std::initializer_list<SOElement>` constructor would be a worthwhile change.  It feels like a good idea to me at the moment.  There are few enough `SOTemplates` that I believe I could make such a change part of this pull request.  Or I could make a separate pull request...

Reply:
I'm fine with either approach - thanks for clarifying. If you think `initializer_list` makes the error state slightly saner, that sounds fine to me.

Reply:
Consider the change I made in #2863, allowing a function to be passed into the `SOTemplate` constructor:

https://github.com/ripple/rippled/blob/9a4583042d0166c44daca829efe9cc80de19a845/src/ripple/protocol/SOTemplate.h#L80-L92

See it in use here: https://github.com/ripple/rippled/blob/9a4583042d0166c44daca829efe9cc80de19a845/src/ripple/app/misc/impl/Manifest.cpp#L42

I'd _prefer_ that we move to this sort of construction for _all_ `SOTemplate` objects going forward and eschew the piece-meal approach we have now.

Reply:
I like your suggestion, but I also have a different approach which calls the `SOTemplate` constructor with a `std::initializer_list`.  I just about have the `initializer_list` approach coded up and tested.  I'm going to submit it as a separate pull request because the changes are too extensive to just slide on top of this pull request.  With luck I'll have that pull request ready today or tomorrow.

Once you see the `initializer_list` approach we can talk about whether we prefer that or the callback.  My feeling is we should only use one approach or the other, and I personally don't have a strong preference at the moment.